nitro_rpc/src/NitroRPC.sol

```diff
@@ -1,23 +1,82 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.17;
 
+
+import {INitroTypes} from "nitro/interfaces/INitroTypes.sol";
+import {NitroUtils} from "nitro/libraries/NitroUtils.sol";
+import {IForceMoveApp} from "nitro/interfaces/IForceMoveApp.sol";
+
 /**
  * @title NitroRPC
  * @dev Solidity implementation of Nitro RPC protocol structures
  */
-interface NitroRPC {
-
+contract NitroRPC is IForceMoveApp {
     struct Payload {
         uint256 requestId;
+        uint256 timestamp;
         string method;
         bytes params;
         bytes result;
-        uint256 timestamp;
     }
 
     struct PayloadSigned {
       Payload rpcMessage;
-      INitroTypes.Signature clientSig
-      INitroTypes.Signature serverSig
+      INitroTypes.Signature clientSig;
+      INitroTypes.Signature serverSig;
+    }
+
+    enum AllocationIndices {
+        Client,
+        Server
+    }
+
+    /**
+     * @notice Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @dev Encodes application-specific rules for a particular ForceMove-compliant state channel.
+     * @param fixedPart Fixed part of the state channel.
+     * @param proof Array of recovered variable parts which constitutes a support proof for the candidate.
+     * @param candidate Recovered variable part the proof was supplied for.
+     */
+    function stateIsSupported(
+        INitroTypes.FixedPart calldata fixedPart,
+        INitroTypes.RecoveredVariablePart[] calldata proof,
+        INitroTypes.RecoveredVariablePart calldata candidate
+    ) external pure override returns (bool, string memory) {
+        require(fixedPart.participants.length == uint256(AllocationIndices.Server) + 1, "bad number of participants");
+        
+        PayloadSigned memory payloadSigned = abi.decode(candidate.variablePart.appData, (PayloadSigned));
+        requireValidPayload(fixedPart, payloadSigned);
+
+        return (true, "");
+    }
+
+    function requireValidPayload(INitroTypes.FixedPart calldata fixedPart, PayloadSigned memory payloadSigned) internal pure {
+        require(recoverPayloadSigner(payloadSigned.rpcMessage, payloadSigned.clientSig) == fixedPart.participants[uint256(AllocationIndices.Client)], "bad client signature");
+        require(recoverPayloadSigner(payloadSigned.rpcMessage, payloadSigned.serverSig) == fixedPart.participants[uint256(AllocationIndices.Server)], "bad server signature");
+
+        // TODO: verify timestamp and requestId
+    }
+
+    // This pure internal function recovers the signer address from the payload and its signature.
+    function recoverPayloadSigner(Payload memory payload, INitroTypes.Signature memory signature) internal pure returns (address) {
```

Comment:
You can substitute this function with a shorter gist:

```solidity
address recoveredAddress = ECDSA.recover(MessageHashUtils.toEthSignedMessageHash(keccak256(abi.encode(payload))), signature);
```

Or an even shorter and more explicit code with using `using` (ba dum tss) expression:

```solidity
contract NitroRPC {
   using ECDSA for bytes32;
   using MessageHashUtils for bytes32;

   address recovered = keccak256(abi.encode(payload)).toEthSignedMessageHash().recover(signature);

}
```

Reply:
P.S. Import `ECDSA` and `MessageHashUtils` from `openzeppelin/contracts`.